x-pack/plugins/apm/server/routes/agent_explorer/get_agents_items.ts

```diff
@@ -0,0 +1,162 @@
+            {
+              exists: {
+                field: AGENT_NAME,
+              },
+            },
+            {
+              exists: {
+                field: AGENT_VERSION,
+              },
+            },
+            {
+              exists: {
+                field: SERVICE_NODE_NAME,
+              },
+            },
+            ...rangeQuery(start, end),
+            ...environmentQuery(environment),
+            ...kqlQuery(kuery),
+            ...(serviceName ? termQuery(SERVICE_NAME, serviceName) : []),
+            ...(agentLanguage
+              ? termQuery(SERVICE_LANGUAGE_NAME, agentLanguage)
+              : []),
+          ],
+        },
+      },
+      aggs: {
+        sample: {
+          random_sampler: randomSampler,
+          aggs: {
+            services: {
+              terms: {
+                field: SERVICE_NAME,
+                size: MAX_NUMBER_OF_SERVICES,
+              },
+              aggs: {
+                instances: {
+                  cardinality: {
+                    field: SERVICE_NODE_NAME,
+                  },
+                },
+                serviceNodes: {
+                  terms: {
+                    field: SERVICE_NODE_NAME,
+                    size: MAX_NUMBER_OF_SERVICES,
+                  },
+                  aggs: {
+                    sample: {
+                      top_metrics: {
+                        metrics: [
+                          { field: AGENT_NAME } as const,
+                          { field: AGENT_VERSION } as const,
+                          { field: SERVICE_LANGUAGE_NAME } as const,
+                        ],
+                        sort: {
+                          '@timestamp': 'desc' as const,
+                        },
+                      },
+                    },
+                  },
+                },
+                environments: {
+                  terms: {
+                    field: SERVICE_ENVIRONMENT,
+                  },
+                },
+              },
+            },
+          },
+        },
+      },
+    },
+  });
+
+  return (
+    response.aggregations?.sample.services.buckets.map((bucket) => {
+      const agent = bucket.serviceNodes.buckets.reduce(
+        (acc, serviceNode) => ({
+          agentName: serviceNode.sample.top[0].metrics[AGENT_NAME] as AgentName,
+          agentVersion: Array.from(
+            new Set([
+              ...acc.agentVersion,
+              serviceNode.sample.top[0].metrics[AGENT_VERSION] as string,
+            ])
+          ),
+        }),
+        { agentVersion: [] } as {
+          agentName?: AgentName;
+          agentVersion: string[];
+        }
+      );
+
+      return {
+        serviceName: bucket.key as string,
+        environments: bucket.environments.buckets.map(
+          (env) => env.key as string
+        ),
+        agentName: agent.agentName,
+        agentVersion: agent.agentVersion,
+        instances: bucket.instances.value as number,
```

Comment:
I'm using the `exists` in the filter for the `service.node.name` wouldn't that be enough?

Reply:
Ah. Well, you shouldn't, because that means we'll miss data where this field is not set. Can you ask for clarification from the APM Server team whether this field can be not set?

Reply:
The `service.node.name` will be set by the server if either a `container.id` or `host.name` are set, otherwise should set it explicitly by the agent. Probably will be omitted only for RUM.

Reply:
In that case I would suggest removing the filter and changing this to `bucket.instances.value || 0` (and adding a comment explaining why we do that)